torch/distributed/fsdp/flat_param.py

```diff
@@ -1017,18 +1147,297 @@ def unflatten_as_params(self) -> Generator:
         flattened parameter, and after the context, restores the original
         parameters as ``Tensor`` views into the flattened parameter.
         """
-        self._unflatten(as_params=True)
+        self._use_unsharded_views(as_params=True)
         try:
             yield
         finally:
-            self._unflatten(as_params=False)
+            self._use_unsharded_views(as_params=False)
+
+    @torch.no_grad()
+    def _use_sharded_views(self) -> None:
+        """
+        Sets the original module parameter variables' data to be flattened
+        views into the sharded flattened parameter.
+
+        The views are kept as flattened to simplify the case where a parameter
+        is sharded across ranks. Parameters whose data is not present in the
+        sharded flattened parameter have their data set to a size-0 empty
+        tensor. We do not delete them to ensure to preserve expected behaviors
+        like model printability. Parameters whose data is present must preserve
+        their variables to be passable to an optimizer.
+        """
+        if not self.uses_sharded_strategy:
+            # For `NO_SHARD`, use the *unflattened* unsharded views since we
+            # have the unsharded parameter
+            self._use_unsharded_views(as_params=True)
+            return
+        self._check_sharded(self.flat_param)
+        start, end = self.flat_param._shard_indices  # type: ignore[attr-defined]
+        offset = 0
+        assert self.flat_param._params is not None
+        for i, (param, (param_name, module, _)) in enumerate(
+            zip(self.flat_param._params, self.flat_param._param_infos)
+        ):
+            setattr(module, param_name, param)
+            in_sharded_flat_param = (
+                i >= start
+                and i <= end
+                and self.flat_param._shard_param_offsets  # type: ignore[attr-defined]
```

Comment:
does this final check just check for validity / flat_param._shard_param_offsets not being None?

Reply:
Yes, I think there is an edge case such that `_shard_param_offsets` is empty (when the rank has no part of the `flat_param`).